### DISCUSSION TITLE
Jk bms ble full protocol details

### DISCUSSION THREAD

**Discussion by aquaforum** [syssi/esphome-jk-bms#321] [General]
Hello
  I used [these](https://github.com/syssi/esphome-jk-bms/blob/main/components/jk_bms_ble/jk_bms_ble.cpp) files as a basis and created a program in C++/Qt to communicate with the bms board.
Everything works, but there are only two commands in those files - COMMAND_CELL_INFO and COMMAND_DEVICE_INFO
commands are formed into a stream of bytes by the function bool JkBmsBle::write_register(uint8_t address, uint32_t value, uint8_t length) (line 1224)
I get the answers to them in function void JkBmsBle::decode_(const std::vector<uint8_t> &data) (line 409)
for me, the switch works on frame_type = 0x01 (followed by a call to this->decode_jk02_settings_(data); (line 418) or frame_type = 0x03 followed by a call to this->decode_device_info_(data); (line 429)
For some reason frame_type = 0x02 does not come....
Is there a complete description of this protocol?
This repository is the only thing I found that somehow works.
PS my board is bms B2A8S20P
Thanks.

**Comment by syssi** [comment]
You could try to use this project and enable the log level `VERY_VERBOSE`. This will print the raw BLE communication to the log. There is no official protocol description. The BLE communication was reversed engineered using btsnoop captures.

If you call the `COMMAND_CELL_INFO` the BMS starts to publish one cell info frame (BLE notification) per second.

**Comment by aquaforum** [comment]
Hello.
My modified program seems to work.
I am attaching screenshots and code. The code is very fresh, I'm working on it.
I decoded the first 200 bytes, the last hundred bytes - no, maybe you can help with this.
The decoding code itself is in void MainWidget::decode_jk02_cell_info_(const QByteArray &data) (line 286)
[JkBmsBle.zip](https://github.com/syssi/esphome-jk-bms/files/11358806/JkBmsBle.zip)
![Screenshot_20230429_121954](https://user-images.githubusercontent.com/807188/235295770-e3c3b045-6f56-4558-a7ac-b93e7b44c62c.png)
![Screenshot_20230429_121806](https://user-images.githubusercontent.com/807188/235295774-a82fc01e-f6aa-4d62-a3be-0ee46b6f4b8f.png)
![Screenshot_20230429_121929](https://user-images.githubusercontent.com/807188/235295777-0cdcf1e1-dd1e-4edc-a880-ac8820cda1b3.png)
